### DISCUSSION TITLE
Fetch on an array of pointers

### DISCUSSION THREAD

**Discussion by project-saf** [netreconlab/Parse-Swift#137] [Q&A]
Is there an easy way to convert an `Array<Pointer<ParseObject>>` into an `Array<ParseObject>` without looping over it and calling fetch() of each pointer?

**Comment by cbaker6** [comment]
This all depends on your particular Parse Class setup on your server, but the Swift Playgrounds is the first place to check. If you designed your schema to be [ParseObject] you can use the example in playgrounds https://github.com/netreconlab/Parse-Swift/blob/c0a3a452a20d9f3ab9cb2a2ea790d0359a3beca0/ParseSwift.playground/Pages/8%20-%20Pointers.xcplaygroundpage/Contents.swift#L128-L242

**Comment by project-saf** [comment]
Perhaps I'm misunderstanding, but the playground didn't really help.
However, I _think_ I have the answer, but would appreciate your thoughts.

Say I have a Parse Class called Movie, and one called Actor, as so:
```swift
public struct Movie: ParseObject {
    // ...
    // default variables
    // ...
    public var movieName: String?
    public var leadActor: Pointer<Actor>?
}

public struct Actor: ParseObject {
    // ...
    // default variables
    // ...
    public var actorName: String?
    public var mostFamousMovie: Pointer<Movie>?
}
```

If I ran a query on the Movie table, for example, where `movieName` contains "James Bond", I'll get an array of Movies. 
If I wanted to get a list of the leadActor's, is the best way to do that via getting an array of pointers, and then convert that to an array of objects, and run `fetchAll()`?
For example:
```swift
guard let movies = try? await query.find()
else { return }

let pointers = movies.compactMap(\.leadActor)
let actors: Array<Actor> = pointers.map { $0.toObject() }
// I know I could have also done:
// let actors: Array<Actor> = movies.compactMap { $0.leadActor?.toObject() }
guard let fetchedActors = try? await actors.fetchAll()
else { return }
```
**Is this the best/most efficient way to do this?**



When I first wrote the question, I was thinking I had to loop over the array of Movie, and then call fetch() on each leadActor pointer.
```swift
guard let movies = try? await query.fetch()
else { throw /*some Error*/ }

var actors: Array<Actor> = []
for movie in movies {
    guard let actor = try? await movie.leadActor?.fetch()
    else { continue }
    actors.append(actor)
}
```
Because I noticed there was no `fetchAll()` method on `Array<Pointer<Movie>>`, in the same way that there was one defined on `Array<Movie>`, which is where my question originated from.

**Reply by cbaker6** [reply]
Yes, using compactMap toObject and then fetchAll is fine. If you submit a PR using map toObject to enable fetchAll for pointers, I’ll review it

**Comment by cbaker6** [comment]
The playgrounds examples shows you don’t need `Pointer<Actor>`, instead 

public struct Movie: ParseObject {
    // ...
    // default variables
    // ...
    public var movieName: String?
    public var leadActor: Actor?
    public var actors: [Actor]?
}

And then query with fetching all to get everything at once.

**Reply by project-saf** [reply]
Yes, but I wouldn't be able to define Actor with the property `mostFamousMovie` being of type `Movie`, due to recursive referencing.

**Reply by cbaker6** [reply]
My answer to this is same as before https://github.com/netreconlab/Parse-Swift/discussions/109#discussioncomment-5934797

**Reply by project-saf** [reply]
I am aware...that's my entire point.
Don't worry about it

**Comment by project-saf** [comment]
Okay, thanks for your thoughts
